### PR TITLE
chore(cdk-lib): migrate to jsii@5.0 / jsii-rosetta@5.0

### PR DESCRIPTION
**Pull request by RomainMuller**
Updates dependencies & code as necessary to use `jsii@5.0` to build the AWS CDK.

BREAKING CHANGE: The return type of `aws-cdk-lib.aws_ec2.SecurityGroup.determineRuleScope` was changed from a tuple (`[SecurityGroupBase, string]`) to a struct with the same values, because tuple types are not supported over the jsii interoperability layer, but `jsii@v1` was incorrectly allowing this to be represented as the `JSON` primitive type. This made the API unusable in non-JS languages. The type of the `metadata` property of `aws-cdk-lib.aws_s3_deployment.BucketDeploymentProps` was changed from an index-only struct to an inline map, because `jsii@v1` silently ignored the index signature (which is otherwise un-supported), resulting in an empty object in non-JS/TS languages. As a consequence, the values of that map can no longer be `undefined` (as `jsii` does not currently support nullable elements in collections).

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/cloud-assembly-schema/lib/cloud-assembly/context-queries.ts [hunk 1]
```diff
@@ -492,6 +492,8 @@ export interface PluginContextQuery {
 
   /**
    * Arbitrary other arguments for the plugin
+   *
+   * @internal
    */
   [key: string]: any;
```
Comment (RomainMuller) [inline packages/@aws-cdk/cloud-assembly-schema/lib/cloud-assembly/context-queries.ts:498]: This is silently ignored by `jsii@v1`, so impossible to use in non-JS/TS languages. I propose to make this `@internal` or to remove it.

**Review by RomainMuller — COMMENTED** [review]
The bulk of code changes here fall into one of these categories:
- `Promise<void>` must be explicitly typed, or TypeScript will require a parameter be provided to the `resolve` handler.
- `catch` bindings are `unknown` by default, explicitly typed as `any` in places where this was assumed, and removed the binding entirely in the few places where it was not used.
- `abstract async` is no longer valid, removed the `async` keyword (the contract really is just "returns a promise", which is expressed fully with the return type declaration)
- Some generic helper methods needed tweaking due to hardened type checking safety around uses of `Object.keys`, `Object.values`, `Object.entries`.
- Some places blindly assumed `Object.keys` returns only `string`, but it could contain `Symbol` objects which have a throwing `toString` (they need to be strigified as `String(sym)`). The 4.9 TypeScript compiler actually reports this as an error. Hardened such uses.
- Adding `overrides` keywords as I am proposing to enable `noImplicitOverride`

Then, there are some more annoying aspects where `jsii@v4.9` checks (and prohibits) things that `v1` silently ignores or mis-interprets (back-porting of these checks to `v1` is planned, but they'll likely need to be WARNING there). I am adding a comment on each use-case to facilitate discussion.

File: packages/cdk-cli-wrapper/lib/cdk-wrapper.ts [hunk 2]
```diff
@@ -66,16 +66,7 @@ export interface SynthFastOptions {
...
-
-/**
- * Additional environment variables to set
- * in the execution environment that will be running
- * the cdk commands
- */
-export interface Environment {
-  [key: string]: string | undefined
+  readonly env?: { [name: string]: string; },
 }
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/cdk-cli-wrapper/lib/cdk-wrapper.ts:70]: Also make jsii throw on signatures like this?

File: packages/@aws-cdk/core/lib/private/asset-staging.ts [hunk 3]
```diff
@@ -211,10 +212,25 @@ export function dockerExec(args: string[], options?: SpawnSyncOptions) {
...
     }
-    throw new Error(`${prog} exited with status ${proc.status}`);
+
+    throw new Error([
+      `${prog} exited with ${reason}`,
+      ...prependLines('--> STDOUT:  ', proc.stdout ) ?? [],
+      ...prependLines('--> STDERR:  ', proc.stderr ) ?? [],
+      `--> Command: ${command}`,
+    ].join('\n'));
   }
```
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/core/lib/private/asset-staging.ts:234]: Yas plz :)

File: packages/cdk-cli-wrapper/lib/cdk-wrapper.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/cdk-cli-wrapper/lib/cdk-wrapper.ts:70]: These changes are motivated 100% by how jsii now throws on such signatures :)

File: packages/@aws-cdk/core/lib/private/asset-staging.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/core/lib/private/asset-staging.ts:234]: https://github.com/aws/aws-cdk/pull/24596

File: packages/@aws-cdk/aws-ec2/lib/security-group.ts [hunk 4]
```diff
@@ -178,15 +178,15 @@ abstract class SecurityGroupBase extends Resource implements ISecurityGroup {
     peer: IPeer,
     connection: Port,
     fromTo: 'from' | 'to',
-    remoteRule?: boolean): [SecurityGroupBase, string] {
+    remoteRule?: boolean): RuleScope {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-ec2/lib/security-group.ts:181]: Note - I could not find any use of this API outside of the CDK Construct library on a cursory search on GitHub.

File: packages/@aws-cdk/aws-apigatewayv2/lib/http/authorizer.ts [hunk 5]
```diff
@@ -248,7 +248,7 @@ export interface IHttpRouteAuthorizer {
   bind(options: HttpRouteAuthorizerBindOptions): HttpRouteAuthorizerConfig;
 }
 
-function undefinedIfNoKeys<A>(obj: A): A | undefined {
+function undefinedIfNoKeys<A extends { [key: string]: unknown }>(obj: A): A | undefined {
```
Comment (Naumel) [inline packages/@aws-cdk/aws-apigatewayv2/lib/http/authorizer.ts:251]: [breadcrumb] Anything is assignable to unknown, but unknown isn't assignable to anything but itself. Diff from `any`.

**Review by Naumel — APPROVED** [review]
Relevant: https://github.com/aws/aws-cdk/pull/24425/files?diff=split&w=1#r1134057642, therefore approving √

I am marking the PR as `pr/do-not-merge` to allow others to review, if they chose to do so.

**Pull request merged**